pkg/policy/repository.go

```diff
@@ -0,0 +1,198 @@
+// Copyright 2016-2017 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package policy
+
+import (
+	"encoding/json"
+	"strings"
+	"sync"
+
+	"github.com/cilium/cilium/pkg/labels"
+	"github.com/cilium/cilium/pkg/policy/api"
+)
+
+// Repository is a list of policy rules which in combination form the security
+// policy. A policy repository can be
+type Repository struct {
+	// Mutex protects the whole policy tree
+	Mutex sync.RWMutex
+	rules []*rule
+}
+
+// NewPolicyRepository allocates a new policy repository
+func NewPolicyRepository() *Repository {
+	return &Repository{}
+}
+
+// CanReachRLocked evaluates the policy repository for the provided search
+// context and returns the verdict or api.Undecided if no rule matches. The
+// policy repository mutex must be held.
+func (p *Repository) CanReachRLocked(ctx *SearchContext) api.Decision {
+	decision := api.Undecided
+
+	for _, r := range p.rules {
+		switch r.canReach(ctx) {
+		// The rule contained a constraint which was not met, this
+		// connection is not allowed
+		case api.Denied:
```

Comment:
`api.Denied` is only returned if it is denied for sure. This only happens for Require rules. If the requirement for a require rule is not met, the potential consumer is never allowed.